cli/tools/test_runner.rs

```diff
@@ -346,6 +346,181 @@ pub async fn run_test_file(
   Ok(())
 }
 
+fn programs_from_source_comments(
+  program_state: Arc<ProgramState>,
+  file: &File,
+) -> Result<Vec<ModuleSpecifier>, AnyError> {
+  let mut specifiers = Vec::new();
+
+  let parsed_module =
+    ast::parse(&file.specifier.as_str(), &file.source, &file.media_type)?;
+
+  let mut comments = parsed_module.get_comments();
+  comments.sort_by_key(|comment| {
+    let location = parsed_module.get_location(&comment.span);
+    location.line
+  });
+
+  let blocks_regex = Regex::new(r"```([^\n]*)\n([\S\s]*?)```")?;
+  let lines_regex = Regex::new(r"(?:\* ?)(?:\# ?)?(.*)")?;
+
+  for comment in comments {
+    if comment.kind != CommentKind::Block || !comment.text.starts_with('*') {
+      continue;
+    }
+
+    for block in blocks_regex.captures_iter(&comment.text) {
+      let maybe_attributes = block.get(1).map(|m| m.as_str().split(' '));
+      let media_type = if let Some(mut attributes) = maybe_attributes {
+        match attributes.next() {
+          Some("js") => MediaType::JavaScript,
+          Some("jsx") => MediaType::Jsx,
+          Some("ts") => MediaType::TypeScript,
+          Some("tsx") => MediaType::Tsx,
+          Some("") => file.media_type,
+          _ => MediaType::Unknown,
+        }
+      } else {
+        file.media_type
+      };
+
+      if media_type == MediaType::Unknown {
+        continue;
+      }
```

Comment:
There's a lot of duplication in `programs_from_fenced_blocks`; could you extract some helpers?

Reply:
I've rewritten it to be less awful.